File: cplusplus/239SlidingWindowMaximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> k_values;
        vector<int> res;
        
        int i = 0;
        int max = INT_MIN;
        for (; i < nums.size() && i < k; ++i) {
            if (max < nums[i]) {
                max = nums[i];
            }
            
            k_values.push_back(nums[i]);
        }
        
        if (i < k || i == 0) {
            return res;
        }
        
        res.push_back(max);
        
        for (; i < nums.size(); ++i) {
            if (nums[i - k] >= max && nums[i - k] > nums[i]) {
                max = INT_MIN;
                for (int j = i - k + 1; j <= i; ++j)
                    if (max < nums[j])
                        max = nums[j];              
            } else if (max < nums[i]) {
                max = nums[i];
            }
            
            res.push_back(max);
        }
        
        return res;
    }
};

File: java/239SlidingWindowMaximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        if (nums == null || nums.length < k || k <= 0)
            return new int[0];
        
        TreeMap<Integer, Integer> map = new TreeMap<>();
        for (int i = 0; i < k - 1; ++i) {
            map.put(nums[i], map.getOrDefault(nums[i], 0) + 1);
        }
        
        int[] res = new int[nums.length - k + 1];
        for (int i = k - 1; i < nums.length; ++i) {
            map.put(nums[i], map.getOrDefault(nums[i], 0) + 1);
            res[i - k + 1] = map.lastKey();
            
            if (map.get(nums[i - k + 1]) == 1)
                map.remove(nums[i - k + 1]);
            else map.put(nums[i - k + 1], map.get(nums[i - k + 1]) - 1);
        }
        
        return res;
    }
}
